XUtils.c

```diff
@@ -210,3 +211,46 @@ char* xStrndup(const char* str, size_t len) {
    }
    return data;
 }
+
+static ssize_t readfd_internal(int fd, void* buffer, size_t count) {
+   ssize_t alreadyRead = 0;
+   count--; // reserve one for null-terminator
+   for (;;) {
+      ssize_t res = read(fd, buffer, count);
+      if (res == -1) {
+         if (errno == EINTR)
+            continue;
+
+         close(fd);
+         return -errno;
+      }
+
+      if (res > 0) {
+         buffer = ((char*)buffer) + res;
+         count -= (size_t)res;
+         alreadyRead += res;
+      }
+
+      if (count == 0 || res == 0) {
+         close(fd);
+         ((char*)buffer)[alreadyRead] = '\0';
```

Comment:
Avoid buffer overflow:
```suggestion
         if (count)
            ((char*)buffer)[alreadyRead] = '\0';
```

Reply:
there is previously the line `count--; // reserve one for null-terminator`.

Or should we check
```c
ssize_t xReadfile(const char* pathname, void* buffer, size_t count) {
   if (count == 0)
      return -EINVAL;
   ...
```
?

Reply:
Unconditionally decrementing causes `MAX_INT` to be read, which is likely undesired as well. But decrementing the `count` on entry is quite a good solution. So return `-EINVAL` on `!count` before decrementing it. Didn't notice the decrement when I initially read through the source.